x-pack/solutions/security/plugins/security_solution/public/cases/components/ai_for_soc/wrapper.test.tsx

```diff
@@ -0,0 +1,161 @@
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import React from 'react';
+import { act, render } from '@testing-library/react';
+import {
+  AiForSOCAlertsTable,
+  CONTENT_TEST_ID,
+  ERROR_TEST_ID,
+  LOADING_PROMPT_TEST_ID,
+  SKELETON_TEST_ID,
+} from './wrapper';
+import { useKibana } from '../../../common/lib/kibana';
+import { TestProviders } from '../../../common/mock';
+import { useFetchIntegrations } from '../../../detections/hooks/alert_summary/use_fetch_integrations';
+import { useFindRulesQuery } from '../../../detection_engine/rule_management/api/hooks/use_find_rules_query';
+
+jest.mock('./table', () => ({
+  Table: () => <div />,
+}));
+jest.mock('../../../common/lib/kibana');
+jest.mock('../../../detections/hooks/alert_summary/use_fetch_integrations');
+jest.mock('../../../detection_engine/rule_management/api/hooks/use_find_rules_query');
+
+const id = 'id';
+const query = { ids: { values: ['abcdef'] } };
+const onLoaded = jest.fn();
+
+describe('<AiForSOCAlertsTab />', () => {
+  beforeEach(() => {
+    jest.clearAllMocks();
+
+    (useFetchIntegrations as jest.Mock).mockReturnValue({
+      installedPackages: [],
+      isLoading: false,
+    });
+    (useFindRulesQuery as jest.Mock).mockReturnValue({
+      data: [],
+      isLoading: false,
+    });
+  });
+
+  it('should render a loading skeleton while creating the dataView', async () => {
+    (useKibana as jest.Mock).mockReturnValue({
+      services: {
+        data: {
+          dataViews: {
+            create: jest.fn(),
+            clearInstanceCache: jest.fn(),
+          },
+        },
+        http: { basePath: { prepend: jest.fn() } },
+      },
+    });
+
+    await act(async () => {
+      const { getByTestId } = render(
+        <AiForSOCAlertsTable id={id} onLoaded={onLoaded} query={query} />
+      );
+
+      expect(getByTestId(LOADING_PROMPT_TEST_ID)).toBeInTheDocument();
+      expect(getByTestId(SKELETON_TEST_ID)).toBeInTheDocument();
+    });
+  });
+
+  it('should render a loading skeleton while fetching packages (integrations)', async () => {
+    (useKibana as jest.Mock).mockReturnValue({
+      services: {
+        data: {
+          dataViews: {
+            create: jest.fn(),
+            clearInstanceCache: jest.fn(),
+          },
+        },
+        http: { basePath: { prepend: jest.fn() } },
+      },
+    });
+    (useFetchIntegrations as jest.Mock).mockReturnValue({
+      installedPackages: [],
+      isLoading: true,
+    });
+
+    await act(async () => {
+      const { getByTestId } = render(
+        <AiForSOCAlertsTable id={id} onLoaded={onLoaded} query={query} />
+      );
+
+      await new Promise(process.nextTick);
+
+      expect(getByTestId(LOADING_PROMPT_TEST_ID)).toBeInTheDocument();
+      expect(getByTestId(SKELETON_TEST_ID)).toBeInTheDocument();
```

Comment:
You almost never need act() or messing with new Promise(process.nextTick) when dealing with render()

There are already rtl builtins like (await findBy... or waitFor(() => {})) which internally use act() and wait as much as needed for all your updates to pass until your condition is met.

This is way more declarative and idiomatic. 

```suggestion
    render(
        <AiForSOCAlertsTable id={id} onLoaded={onLoaded} query={query} />
      );

      expect(await screen.findByTestId(LOADING_PROMPT_TEST_ID)).toBeInTheDocument();
      expect(await screen.findByTestId(SKELETON_TEST_ID)).toBeInTheDocument();
```

And it passes

![image](https://github.com/user-attachments/assets/69dde78d-3deb-4d30-b165-4a0fce877a27)

Reply:
How would you do it for the first test?

Reply:
I changed the tests except the first one

Reply:
sorry for late response I zoom out sometimes, the first hits the second category I outlined above and passes perfectly with waitFor

```
    render(<AiForSOCAlertsTable id={id} onLoaded={onLoaded} query={query} />);

    await waitFor(() => {
      expect(screen.getByTestId(LOADING_PROMPT_TEST_ID)).toBeInTheDocument();
      expect(screen.getByTestId(SKELETON_TEST_ID)).toBeInTheDocument();
    });
```

But btw seems like in the source code you are wrapping EuiSkeletonRectangle in EuiSkeletonLoading, which is unnecessary. EuiSkeletonRectangle is already using EuiSkeletonLoading underneath and you can pass both isLoading prop into it and pass children that will be used for loadedContent. Here look: https://github.com/elastic/eui/blob/8e67c2122721c6a623bacbc60a4e4497888b9941/packages/eui/src/components/skeleton/skeleton_rectangle.tsx#L51

So changing to this works:

```tsx
  return (
    <EuiSkeletonRectangle
      isLoading={integrationIsLoading || dataViewLoading}
      data-test-subj={SKELETON_TEST_ID}
      height={400}
      width="100%"
    >
      <>
        {!dataView || !dataView.id ? (
          <EuiEmptyPrompt
            color="danger"
            data-test-subj={ERROR_TEST_ID}
            iconType="error"
            title={<h2>{DATAVIEW_ERROR}</h2>}
          />
        ) : (
          <div data-test-subj={CONTENT_TEST_ID}>
            <Table
              dataView={dataView}
              id={id}
              onLoaded={onLoaded}
              packages={installedPackages}
              query={query}
              ruleResponse={ruleResponse}
            />
          </div>
        )}
      </>
    </EuiSkeletonRectangle>
  );

```

And this means you no longer need LOADING_PROMPT_TEST_ID locator in source and tests which was apparently a duplication.

Reply:
Good point, this is because originally I did have more skeleton items, but now that I have only a single rectangle this is cleaner, thanks!

Reply:
Fixed!